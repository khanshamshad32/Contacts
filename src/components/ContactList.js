import React from 'react';
import {
  FlatList,
  Text,
  View,
  Image,
  StyleSheet,
  TouchableOpacity,
} from 'react-native';
import {connect} from 'react-redux';
import * as Colors from '../utils/color';

export const ContactList = ({navigation, contacts, favourite}) => {
  const renderItem = ({item}) => {
    return (
      <TouchableOpacity
        style={Styles.itemContainerStyle}
        activeOpacity={0.8}
        onPress={() => {
          navigation.navigate('ContactUpsert', {
            contact: item,
          });
        }}>
        <View style={Styles.itemImageContainerStyle}>
          <Image
            source={{uri: item.avatar || 'ic_account'}}
            resizeMode="cover"
            style={Styles.itemImageStyle}
          />
        </View>
        <Text style={Styles.itemNameStyle}>{item.name}</Text>
      </TouchableOpacity>
    );
  };

  // console.log('Contact List Render', contacts);

  return (
    <>
      <FlatList
        style={Styles.rootContainerStyle}
        data={contacts}
        renderItem={renderItem}
        keyExtractor={item => `${item.id}`}
        extraData={contacts}
        bounces={true}
        bouncesZoom={true}
      />
      {!favourite ? (
        <TouchableOpacity
          style={Styles.addButtonStyle}
          activeOpacity={0.8}
          onPress={() => {
            navigation.navigate('ContactUpsert');
          }}>
          <Image
            style={Styles.addButtonImageStyle}
            source={{uri: 'ic_add'}}
            resizeMode="contain"
          />
        </TouchableOpacity>
      ) : null}
    </>
  );
};

export const mapStateToProps = ({contacts}, {favourite}) => {
  // console.log('mapStateToProps contacts', contacts);

  if (favourite) {
    contacts = contacts.filter(item => item.favourite);
  }

  return {contacts};
};

const Styles = StyleSheet.create({
  rootContainerStyle: {
    backgroundColor: Colors.APP_WHITE,
  },
  itemContainerStyle: {
    flexDirection: 'row',
    alignItems: 'center',
    padding: 10,
    height: 80,
    marginVertical: 3,
    marginLeft: 30,
  },
  itemImageContainerStyle: {
    height: 50,
    aspectRatio: 1,
    overflow: 'hidden',
    borderRadius: 25,
    borderWidth: 1,
    borderColor: Colors.APP_GREY,
  },
  itemImageStyle: {
    flex: 1,
  },
  itemNameStyle: {
    marginLeft: 20,
    fontSize: 18,
    fontWeight: '600',
    color: Colors.BLACK_ACTIVE,
  },
  addButtonStyle: {
    position: 'absolute',
    right: 30,
    bottom: 40,
    width: 50,
    aspectRatio: 1,
    borderRadius: 25,
    overflow: 'hidden',
  },
  addButtonImageStyle: {
    flex: 1,
  },
});

export default connect(mapStateToProps)(ContactList);
