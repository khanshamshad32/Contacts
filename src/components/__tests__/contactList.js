/* eslint-disable prettier/prettier */
import React from 'react';
import renderer from 'react-test-renderer';
import {ContactList, mapStateToProps} from '../ContactList';

const favouriteContacts = [
  {name: 'Ola', id: 3, favourite: true},
  {name: 'Erick', id: 4, favourite: true},
];

const defaultContacts = [
  {name: 'Jack', id: 1},
  {name: 'Zoe', id: 2},
  ...favouriteContacts,
];

describe('Contact List', () => {
  it('mapStateToProp should return all contacts', () => {
    let state = {contacts: [...defaultContacts]};
    expect(mapStateToProps(state, {}).contacts).toEqual(defaultContacts);
  });

  it('should render correctly', () => {
    let tree = renderer.create(<ContactList contacts={defaultContacts} />);
    expect(tree.toJSON()).toMatchSnapshot();
  });
});

describe('Favourite Contact List', () => {
  it('mapStateToProp should return only favourite contacts', () => {
    let state = {contacts: [...defaultContacts]};
    expect(mapStateToProps(state, {favourite: true}).contacts).toEqual(
      favouriteContacts,
    );
  });

  it('should render correctly', () => {
    let tree = renderer.create(
      <ContactList contacts={favouriteContacts} favourite />,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
