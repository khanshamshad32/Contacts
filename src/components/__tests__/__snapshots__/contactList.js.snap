// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact List should render correctly 1`] = `
Array [
  <RCTScrollView
    bounces={true}
    bouncesZoom={true}
    data={
      Array [
        Object {
          "id": 1,
          "name": "Jack",
        },
        Object {
          "id": 2,
          "name": "Zoe",
        },
        Object {
          "favourite": true,
          "id": 3,
          "name": "Ola",
        },
        Object {
          "favourite": true,
          "id": 4,
          "name": "Erick",
        },
      ]
    }
    disableVirtualization={false}
    extraData={
      Array [
        Object {
          "id": 1,
          "name": "Jack",
        },
        Object {
          "id": 2,
          "name": "Zoe",
        },
        Object {
          "favourite": true,
          "id": 3,
          "name": "Ola",
        },
        Object {
          "favourite": true,
          "id": 4,
          "name": "Erick",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "backgroundColor": "#fff",
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 80,
              "marginLeft": 30,
              "marginVertical": 3,
              "opacity": 1,
              "padding": 10,
            }
          }
        >
          <View
            style={
              Object {
                "aspectRatio": 1,
                "borderColor": "grey",
                "borderRadius": 25,
                "borderWidth": 1,
                "height": 50,
                "overflow": "hidden",
              }
            }
          >
            <Image
              resizeMode="cover"
              source={
                Object {
                  "uri": "ic_account",
                }
              }
              style={
                Object {
                  "flex": 1,
                }
              }
            />
          </View>
          <Text
            style={
              Object {
                "color": "rgba(0,0,0,0.87)",
                "fontSize": 18,
                "fontWeight": "600",
                "marginLeft": 20,
              }
            }
          >
            Jack
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 80,
              "marginLeft": 30,
              "marginVertical": 3,
              "opacity": 1,
              "padding": 10,
            }
          }
        >
          <View
            style={
              Object {
                "aspectRatio": 1,
                "borderColor": "grey",
                "borderRadius": 25,
                "borderWidth": 1,
                "height": 50,
                "overflow": "hidden",
              }
            }
          >
            <Image
              resizeMode="cover"
              source={
                Object {
                  "uri": "ic_account",
                }
              }
              style={
                Object {
                  "flex": 1,
                }
              }
            />
          </View>
          <Text
            style={
              Object {
                "color": "rgba(0,0,0,0.87)",
                "fontSize": 18,
                "fontWeight": "600",
                "marginLeft": 20,
              }
            }
          >
            Zoe
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 80,
              "marginLeft": 30,
              "marginVertical": 3,
              "opacity": 1,
              "padding": 10,
            }
          }
        >
          <View
            style={
              Object {
                "aspectRatio": 1,
                "borderColor": "grey",
                "borderRadius": 25,
                "borderWidth": 1,
                "height": 50,
                "overflow": "hidden",
              }
            }
          >
            <Image
              resizeMode="cover"
              source={
                Object {
                  "uri": "ic_account",
                }
              }
              style={
                Object {
                  "flex": 1,
                }
              }
            />
          </View>
          <Text
            style={
              Object {
                "color": "rgba(0,0,0,0.87)",
                "fontSize": 18,
                "fontWeight": "600",
                "marginLeft": 20,
              }
            }
          >
            Ola
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 80,
              "marginLeft": 30,
              "marginVertical": 3,
              "opacity": 1,
              "padding": 10,
            }
          }
        >
          <View
            style={
              Object {
                "aspectRatio": 1,
                "borderColor": "grey",
                "borderRadius": 25,
                "borderWidth": 1,
                "height": 50,
                "overflow": "hidden",
              }
            }
          >
            <Image
              resizeMode="cover"
              source={
                Object {
                  "uri": "ic_account",
                }
              }
              style={
                Object {
                  "flex": 1,
                }
              }
            />
          </View>
          <Text
            style={
              Object {
                "color": "rgba(0,0,0,0.87)",
                "fontSize": 18,
                "fontWeight": "600",
                "marginLeft": 20,
              }
            }
          >
            Erick
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>,
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "aspectRatio": 1,
        "borderRadius": 25,
        "bottom": 40,
        "opacity": 1,
        "overflow": "hidden",
        "position": "absolute",
        "right": 30,
        "width": 50,
      }
    }
  >
    <Image
      resizeMode="contain"
      source={
        Object {
          "uri": "ic_add",
        }
      }
      style={
        Object {
          "flex": 1,
        }
      }
    />
  </View>,
]
`;

exports[`Favourite Contact List should render correctly 1`] = `
<RCTScrollView
  bounces={true}
  bouncesZoom={true}
  data={
    Array [
      Object {
        "favourite": true,
        "id": 3,
        "name": "Ola",
      },
      Object {
        "favourite": true,
        "id": 4,
        "name": "Erick",
      },
    ]
  }
  disableVirtualization={false}
  extraData={
    Array [
      Object {
        "favourite": true,
        "id": 3,
        "name": "Ola",
      },
      Object {
        "favourite": true,
        "id": 4,
        "name": "Erick",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Object {
      "backgroundColor": "#fff",
    }
  }
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 80,
            "marginLeft": 30,
            "marginVertical": 3,
            "opacity": 1,
            "padding": 10,
          }
        }
      >
        <View
          style={
            Object {
              "aspectRatio": 1,
              "borderColor": "grey",
              "borderRadius": 25,
              "borderWidth": 1,
              "height": 50,
              "overflow": "hidden",
            }
          }
        >
          <Image
            resizeMode="cover"
            source={
              Object {
                "uri": "ic_account",
              }
            }
            style={
              Object {
                "flex": 1,
              }
            }
          />
        </View>
        <Text
          style={
            Object {
              "color": "rgba(0,0,0,0.87)",
              "fontSize": 18,
              "fontWeight": "600",
              "marginLeft": 20,
            }
          }
        >
          Ola
        </Text>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 80,
            "marginLeft": 30,
            "marginVertical": 3,
            "opacity": 1,
            "padding": 10,
          }
        }
      >
        <View
          style={
            Object {
              "aspectRatio": 1,
              "borderColor": "grey",
              "borderRadius": 25,
              "borderWidth": 1,
              "height": 50,
              "overflow": "hidden",
            }
          }
        >
          <Image
            resizeMode="cover"
            source={
              Object {
                "uri": "ic_account",
              }
            }
            style={
              Object {
                "flex": 1,
              }
            }
          />
        </View>
        <Text
          style={
            Object {
              "color": "rgba(0,0,0,0.87)",
              "fontSize": 18,
              "fontWeight": "600",
              "marginLeft": 20,
            }
          }
        >
          Erick
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
